File: js-src/src/storage/repository/VcsAdapter.ts
import { Commit, CommitMetadata } from "../version-control/Commit";
import { CommitGraph } from "../version-control/CommitGraph";

export interface BranchMetadata {
    name: string;
    head_commit_id: string | null;
}

export interface InternalRepoUpdateNoDeltas {
    addedCommits: CommitMetadata[];
    removedCommits: CommitMetadata[];
    updatedCommits: CommitMetadata[];
    addedBranches: BranchMetadata[];
    updatedBranches: BranchMetadata[];
    removedBranches: BranchMetadata[];
}

export interface InternalRepoUpdate extends Omit<InternalRepoUpdateNoDeltas, 'addedCommits' | 'updatedCommits'> {
    addedCommits: Commit[];
    updatedCommits: Commit[];
}

export interface VcsAdapter {
    getCommitGraph(): Promise<CommitGraph>;
    getCommits(ids: string[]): Promise<Commit[]>;
    applyUpdate(update: InternalRepoUpdate): Promise<void>;
    close(): void;
    eraseStorage(): Promise<void>;
}

File: py-src/fusion/storage/vcs_adapter.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any

from fusion.storage.commit import Commit, CommitMetadata
from fusion.storage.commit_graph import BranchMetadata, CommitGraph


@dataclass
class InternalRepoUpdate:
    added_commits: list[Commit]
    removed_commits: list[CommitMetadata]
    updated_commits: list[Commit]
    added_branches: list[BranchMetadata]
    updated_branches: list[BranchMetadata]
    removed_branches: list[BranchMetadata]


class VcsAdapter(ABC):
    @abstractmethod
    def get_commit_graph(self) -> CommitGraph: ...

    @abstractmethod
    def get_commits(self, ids: list[str]) -> list[Commit]: ...

    @abstractmethod
    def apply_update(self, update: InternalRepoUpdate) -> None: ...

    @abstractmethod
    def close(self) -> None: ...

    @abstractmethod
    def erase_storage(self) -> None: ...
